File: examples/example_complex.lua
local string = require("../../rustscript/stdlib/rss/strings.rss")
local io = require("io")
local re = require("re")
local json = require("json")
local runtime = require("runtime")

-- Complex Lua flavor example: loop + stdlib + host + closure.
local total = 0
for i = 0, 3, 1 do
    total = total + i
end

if not string.non_empty("lua") then
    total = 0
else
    total = total + 1
end

local base = 7
local add = function(value) return value + base end
base = 8
local closure_value = add(5)

local profile = { stats = { score = closure_value } }
local chained_score = profile?.stats?.score
local missing_score = profile?.missing?.value

local function pair_scores()
    return closure_value, total
end
local first_only = pair_scores()
local first_score, second_score, third_score = pair_scores()

local function branch_score()
    if true then
        return closure_value
    else
        return closure_value, total
    end
end
local branch_a, branch_b, branch_c = branch_score()

local function keep(value)
    return value
end
local regex_ok = re.match("^lua$", "LUA", "i")
local payload = { lang = "lua", score = closure_value, chained = closure_value }
local payload_json = json.encode(payload)
local payload_decoded = json.decode(payload_json)
local json_score = payload_decoded.score
local sleep_ok = runtime.sleep(100)
local io_ok = true
if true then
    io_ok = io.exists(".")
end

local unpack_ok = first_only == closure_value and first_score == closure_value and second_score == total and third_score == nil and branch_a == closure_value and branch_b == nil and branch_c == nil

if chained_score ~= nil then
    if regex_ok and io_ok and sleep_ok and json_score == closure_value and unpack_ok and missing_score == nil then
        print(keep(chained_score))
    else
        print(0)
    end
else
    print(0)
end

File: examples/example_complex.js
import * as string from "../../rustscript/stdlib/rss/strings.rss";
import * as io from "io";
import * as re from "re";
import * as json from "json";
import * as runtime from "runtime";

// Complex JavaScript flavor example: loop + stdlib + host + closure.
let total = 0;
for (let i = 0; i < 4; i = i + 1) {
    total = total + i;
}

if (!string.non_empty("javascript")) {
    total = 0;
} else {
    total = total + 1;
}

let base = 7;
let add = (value) => value + base;
base = 8;
let closureValue = add(5);

const profile = { stats: { score: closureValue } };
const chainedScoreOpt = profile?.stats?.score;
const missingScore = profile?.missing?.value;

function keep(value) { return value; }
const regexOk = re.match("^javascript$", "JAVASCRIPT", "i");
const payload = {
    lang: "javascript",
    score: closureValue,
    chained: closureValue,
};
const payloadJson = json.encode(payload);
const payloadDecoded = json.decode(payloadJson);
const jsonScore = payloadDecoded.score;
const sleepOk = runtime.sleep(100);
let ioOk = true;
if (true) {
    ioOk = io.exists(".");
}

if (chainedScoreOpt != null) {
    if (regexOk && ioOk && sleepOk && jsonScore == closureValue && missingScore == null) {
        console.log(keep(chainedScoreOpt));
    } else {
        console.log(0);
    }
} else {
    console.log(0);
}
